src/genetic.c

```diff
@@ -35,39 +43,41 @@ static void fillTable(uint16_t* population) {
 static void tourney(float* populationStrength, unsigned int* firstParentIndex,
                     unsigned int* secondParentIndex) {
 
-  float chosenIndexes[populationSize];
-  float bestFitness, secondbestFitness = 0;
+  float chosenIndexes[tournamentSelectionsSize];
+  float bestFitness = FLT_MAX;
+  float secondbestFitness = FLT_MAX;
 
   for (unsigned int i = 0; i < tournamentSelectionsSize; i++) {
 
     unsigned int index =
-        (unsigned int)(linear_congruential_random_generator() * populationSize);
-    unsigned int isAlreadyChosen = 1;
-
-    chosenIndexes[i] = index;
-
-    if (i != 0) {
+        linear_congruential_random_generator() * populationSize;
+    uint8_t isNotAlreadyChosen = 1;
 
-      for (unsigned int j = 0; j < i; j++) {
+    for (unsigned int j = 0; j < i; j++) {
 
-        if (chosenIndexes[j] == index) {
-          i--;
-          isAlreadyChosen = 0;
-          break;
-        }
+      if (chosenIndexes[j] == index) {
+        i--;
+        isNotAlreadyChosen = 0;
+        break;
       }
     }
 
     const float fitness = populationStrength[index];
 
-    if (!i || (isAlreadyChosen && bestFitness > fitness)) {
+    if (isNotAlreadyChosen) {
+      chosenIndexes[i] = index;
+    } else {
+      continue;
+    }
+
+    if (isNotAlreadyChosen && bestFitness > fitness) {
```

Comment:
You can remove the isNotAlreadyChosen since we've validated that the value was true at line 69